File: libs/python/mail_action_usps/notify.py
#!/usr/bin/env python3
"""
Notification routing for USPS mail alerts.

Routes notifications to different recipients based on the addressee:
  - Nicole / Eastside Improv → Nicole's target
  - Jeff / Jeffrey / Steinbok / default → Jeff's target

Config lives at ~/.openclaw/agents/mail/workspace/usps-mail/config.json.
Notifications are planned first, then optionally delivered via `openclaw message send`.
"""

import json
import os
import subprocess
import sys

from .paths import get_config_file

# Addressee patterns → routing key (matched case-insensitively)
_NICOLE_PATTERNS = {"nicole", "eastside improv"}


def load_config(workspace_agent: str = None) -> dict:
    """Load plugin config (routing, etc.)."""
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    config_file = get_config_file(workspace_agent)
    if config_file.exists():
        with open(config_file) as f:
            return json.load(f)
    return {}


def _classify_recipient(addressee: str) -> str:
    """Determine routing key from addressee name."""
    low = (addressee or "").lower()

    for pat in _NICOLE_PATTERNS:
        if pat in low:
            # Joint mail ("Jeffrey & Nicole") → jeff
            if any(p in low for p in ("jeff", "jeffrey")):
                return "jeff"
            return "nicole"

    return "jeff"


def send_message(message: str, channel: str, target: str) -> bool:
    """Send a notification via openclaw message send."""
    if not target:
        print(f"[NOTIFY] no target: {message}", file=sys.stderr)
        return False

    try:
        result = subprocess.run(
            ["openclaw", "message", "send",
             "--channel", channel, "--target", target,
             "--message", message],
            timeout=30, capture_output=True, text=True,
        )
        return result.returncode == 0
    except (FileNotFoundError, subprocess.TimeoutExpired):
        print(f"[NOTIFY] send failed: {message[:100]}", file=sys.stderr)
        return False


def build_notification_plan(date_str: str, items: list, config: dict | None = None, workspace_agent: str = None) -> list:
    """Build per-recipient USPS notification payloads without sending them."""
    if not config and not workspace_agent:
        raise ValueError("workspace_agent is required when config is not provided")
    config = config or load_config(workspace_agent)
    routing = config.get("routing", {})
    default_channel = "discord"

    if not routing:
        # Fallback: use env vars
        routing = {
            "default": {
                "channel": os.environ.get("NOTIFY_CHANNEL", "discord"),
                "target": os.environ.get("NOTIFY_TARGET", ""),
            }
        }

    # Bucket items by recipient
    buckets = {}
    for item in items:
        key = _classify_recipient(item.get("addressee", ""))
        buckets.setdefault(key, []).append(item)

    results = []
    other_items = [it for it in items if it.get("importance") in ("low", "junk", "ad", "medium")]

    for key, key_items in buckets.items():
        dest = routing.get(key) or routing.get("default")
        if not dest:
            continue

        notify_items = [it for it in key_items if it.get("importance") in ("urgent", "high")]
        if not notify_items:
            continue

        lines = [f"📬 USPS Mail ({date_str}) — {len(notify_items)} important for you:"]
        for item in notify_items:
            sender = item.get("sender", "Unknown")
            desc = item.get("description", "")
            imp = item.get("importance", "").upper()
            line = f"  🔴 [{imp}] {sender}"
            if desc:
                line += f": {desc}"
            lines.append(line)

        # Junk/routine summary for Jeff only
        if key != "nicole" and other_items:
            junk = sum(1 for it in other_items if it.get("importance") in ("junk", "ad"))
            rest = len(other_items) - junk
            parts = []
            if rest:
                parts.append(f"{rest} routine")
            if junk:
                parts.append(f"{junk} junk")
            if parts:
                lines.append(f"  Also: {', '.join(parts)}")

        msg = "\n".join(lines)
        channel = dest.get("channel", default_channel)
        target = dest.get("target", "")
        results.append({
            "recipient": key,
            "target": target,
            "channel": channel,
            "message": msg,
            "items": notify_items,
        })

    return results


def route_and_notify(date_str: str, items: list, dry_run: bool = False, workspace_agent: str = None) -> list:
    """
    Route important items to the right recipients and send notifications.

    Returns list of dicts: [{recipient, message, sent}]
    """
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    plan = build_notification_plan(date_str, items, workspace_agent=workspace_agent)
    results = []
    for entry in plan:
        sent = False
        if dry_run:
            print(f"[DRY RUN → {entry['recipient']}/{entry['target']}]\n{entry['message']}\n", file=sys.stderr)
        else:
            sent = send_message(entry["message"], entry["channel"], entry["target"])

        results.append({**entry, "sent": sent})

    return results

File: libs/ts/mail_action_usps/src/notify.ts
/**
 * Notification routing for USPS mail alerts.
 *
 * Routes notifications to different recipients based on the addressee.
 * Config lives at ~/.openclaw/agents/mail/workspace/usps-mail/config.json.
 * Notifications are planned first, then optionally delivered via `openclaw message send`.
 */

import { existsSync, readFileSync } from "node:fs";
import { execFileSync } from "node:child_process";

import { getConfigFile } from "./paths.js";

const NICOLE_PATTERNS = new Set(["nicole", "eastside improv"]);

/**
 * Load plugin config (routing, etc.).
 */
export function loadConfig(
  workspaceAgent?: string,
): Record<string, unknown> {
  if (!workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const configFile = getConfigFile(workspaceAgent);
  if (existsSync(configFile)) {
    return JSON.parse(readFileSync(configFile, "utf-8"));
  }
  return {};
}

/**
 * Determine routing key from addressee name.
 */
export function classifyRecipient(addressee: string | null | undefined): string {
  const low = (addressee ?? "").toLowerCase();

  for (const pat of NICOLE_PATTERNS) {
    if (low.includes(pat)) {
      // Joint mail ("Jeffrey & Nicole") → jeff
      if (low.includes("jeff") || low.includes("jeffrey")) {
        return "jeff";
      }
      return "nicole";
    }
  }

  return "jeff";
}

/**
 * Send a notification via openclaw message send.
 */
export function sendMessage(
  message: string,
  channel: string,
  target: string,
): boolean {
  if (!target) {
    process.stderr.write(`[NOTIFY] no target: ${message}\n`);
    return false;
  }

  try {
    execFileSync(
      "openclaw",
      ["message", "send", "--channel", channel, "--target", target, "--message", message],
      { timeout: 30000, stdio: ["pipe", "pipe", "pipe"] },
    );
    return true;
  } catch {
    process.stderr.write(
      `[NOTIFY] send failed: ${message.slice(0, 100)}\n`,
    );
    return false;
  }
}

export interface NotificationEntry {
  recipient: string;
  target: string;
  channel: string;
  message: string;
  items: Array<Record<string, unknown>>;
  sent?: boolean;
}

/**
 * Build per-recipient USPS notification payloads without sending them.
 */
export function buildNotificationPlan(
  dateStr: string,
  items: Array<Record<string, unknown>>,
  options?: {
    config?: Record<string, unknown>;
    workspaceAgent?: string;
    /** Fallback notification channel when routing config has no default. */
    defaultChannel?: string;
    /** Fallback notification target when routing config has no default. */
    defaultTarget?: string;
  },
): NotificationEntry[] {
  const opts = options ?? {};
  if (!opts.config && !opts.workspaceAgent) {
    throw new Error(
      "workspace_agent is required when config is not provided",
    );
  }
  const config = opts.config ?? loadConfig(opts.workspaceAgent);
  let routing = config.routing as Record<
    string,
    { channel?: string; target?: string }
  > | undefined;
  const defaultChannel = "discord";

  if (!routing || Object.keys(routing).length === 0) {
    routing = {
      default: {
        channel: opts.defaultChannel ?? defaultChannel,
        target: opts.defaultTarget ?? "",
      },
    };
  }

  // Bucket items by recipient
  const buckets: Record<string, Array<Record<string, unknown>>> = {};
  for (const item of items) {
    const key = classifyRecipient(item.addressee as string);
    if (!buckets[key]) buckets[key] = [];
    buckets[key].push(item);
  }

  const results: NotificationEntry[] = [];
  const otherItems = items.filter((it) =>
    ["low", "junk", "ad", "medium"].includes(it.importance as string),
  );

  for (const [key, keyItems] of Object.entries(buckets)) {
    const dest = routing[key] ?? routing.default;
    if (!dest) continue;

    const notifyItems = keyItems.filter((it) =>
      ["urgent", "high"].includes(it.importance as string),
    );
    if (notifyItems.length === 0) continue;

    const lines: string[] = [
      `📬 USPS Mail (${dateStr}) — ${notifyItems.length} important for you:`,
    ];
    for (const item of notifyItems) {
      const sender = (item.sender as string) ?? "Unknown";
      const desc = (item.description as string) ?? "";
      const imp = ((item.importance as string) ?? "").toUpperCase();
      let line = `  🔴 [${imp}] ${sender}`;
      if (desc) line += `: ${desc}`;
      lines.push(line);
    }

    // Junk/routine summary for Jeff only
    if (key !== "nicole" && otherItems.length > 0) {
      const junk = otherItems.filter((it) =>
        ["junk", "ad"].includes(it.importance as string),
      ).length;
      const rest = otherItems.length - junk;
      const parts: string[] = [];
      if (rest) parts.push(`${rest} routine`);
      if (junk) parts.push(`${junk} junk`);
      if (parts.length > 0) {
        lines.push(`  Also: ${parts.join(", ")}`);
      }
    }

    const msg = lines.join("\n");
    const channel = dest.channel ?? defaultChannel;
    const target = dest.target ?? "";
    results.push({
      recipient: key,
      target,
      channel,
      message: msg,
      items: notifyItems,
    });
  }

  return results;
}

/**
 * Route important items to the right recipients and send notifications.
 */
export function routeAndNotify(
  dateStr: string,
  items: Array<Record<string, unknown>>,
  options?: { dryRun?: boolean; workspaceAgent?: string; defaultChannel?: string; defaultTarget?: string },
): NotificationEntry[] {
  const opts = options ?? {};
  if (!opts.workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const plan = buildNotificationPlan(dateStr, items, {
    workspaceAgent: opts.workspaceAgent,
    defaultChannel: opts.defaultChannel,
    defaultTarget: opts.defaultTarget,
  });
  const results: NotificationEntry[] = [];
  for (const entry of plan) {
    let sent = false;
    if (opts.dryRun) {
      process.stderr.write(
        `[DRY RUN → ${entry.recipient}/${entry.target}]\n${entry.message}\n\n`,
      );
    } else {
      sent = sendMessage(entry.message, entry.channel, entry.target);
    }
    results.push({ ...entry, sent });
  }
  return results;
}
